managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java

```diff
@@ -3815,7 +3815,7 @@ public void removeWaitingCursor(ManagedCursor cursor) {
 
     public void addWaitingCursor(ManagedCursorImpl cursor) {
         if (cursor instanceof NonDurableCursorImpl) {
-            if (cursor.isActive()) {
+            if (cursors.get(cursor.getName()) != null) {
```

Comment:
Is it possible to add a comment to explain this condition?

It seems that the original reason to add this was to fix #22157 in #22191 . It would be great to have the rational explained in a comment.

When looking at the code, it seems that this would always evaluate to true since non durable cursors are always "cached" and part of the `cursors` map. Only when the non durable cursor is deleted, would it be removed from `cursors`.

Is the logic in `addWaitingCursor` useful? Do we have a separate test case for #22191 that proves that it fixes the memory leak?

Reply:
Please check my other comment https://github.com/apache/pulsar/pull/22454/files#r1555192761 . This method "addWaitingCursor" can be removed completely since the other change fixes #22157.

Reply:
Please remove addWaitingCursor method or make it simply `this.waitingCursors.add(cursor);` without any conditions.

Reply:
> Is the logic in addWaitingCursor useful? Do we have a separate test case for #22191 that proves that it fixes the memory leak?

`testAddWaitingCursorsForNonDurable` could help reproduce it

Reply:
thanks, continued in the other thread, https://github.com/apache/pulsar/pull/22454#discussion_r1555263091